frontend/src/components/Tree/Tree.jsx

```diff
@@ -0,0 +1,138 @@
+import React, { useState, useRef, useEffect } from "react";
+import styles from "./Tree.module.css";
+import { select, hierarchy, tree, linkHorizontal } from "d3";
+import ResizeObserver from "resize-observer-polyfill";
+import "./Tree.module.css";
+
+const useResizeObserver = (ref) => {
+  const [dimensions, setDimensions] = useState(null);
+  useEffect(() => {
+    const observeTarget = ref.current;
+    const resizeObserver = new ResizeObserver((entries) => {
+      entries.forEach((entry) => {
+        setDimensions(entry.contentRect);
+      });
+    });
+    resizeObserver.observe(observeTarget);
+    return () => {
+      resizeObserver.unobserve(observeTarget);
+    };
+  }, [ref]);
+  return dimensions;
+};
+
+function usePrevious(value) {
+  const ref = useRef();
+  useEffect(() => {
+    ref.current = value;
+  });
+  return ref.current;
+}
+
+export function Tree({ jsonData, onNodeClick }) {
+  const [currentView, setCurrentView] = useState({});
+  const svgRef = useRef();
+  const wrapperRef = useRef();
+  const dimensions = useResizeObserver(wrapperRef);
+
+  // We save data to see if it changed
+  const previouslyRenderedData = usePrevious(currentView);
+
+  useEffect(() => {
+    setCurrentView(jsonData);
+  }, [jsonData]);
+
+  // Will be called initially and on every data change
+  useEffect(() => {
+    const svg = select(svgRef.current);
+
+    // Use dimensions from useResizeObserver,
+    // but use getBoundingClientRect on initial render
+    // (dimensions are null for the first render)
+    const { width, height } =
+      dimensions || wrapperRef.current.getBoundingClientRect();
+
+    // Transform hierarchical data
+    const root = hierarchy(currentView);
+    const treeLayout = tree().size([height, width - 200]);
+
+    // Creates the links between nodes
+    const linkGenerator = linkHorizontal()
+      .x((link) => link.y)
+      .y((link) => link.x);
+
+    // Enrich hierarchical data with coordinates
+    treeLayout(root);
+
+    // Create the node group, which will hold the nodes and labels
+    const nodeGroup = svg.selectAll(".node-group").data(root.descendants());
+    const nodeGroupEnter = nodeGroup.enter().append("g");
+
+    nodeGroupEnter
+      .merge(nodeGroup)
+      .attr("class", "node-group")
+      .attr("transform", (node) => `translate(${node.y},${node.x})`)
+      .style("cursor", "pointer")
+      .on("click", onNodeClick);
+
+    nodeGroup.exit().remove();
```

Comment:
Wondering what this line is for? I removed it locally and nothing seemed to happen; I printed out what it returns and it seemed to be an empty array 
![image](https://user-images.githubusercontent.com/29286974/104182890-0af1f500-5409-11eb-9389-30e8b3ac63a0.png)

Reply:
Try going to the first course and clicking on "Front". You'll see that it doesn't remove the "Front" node on the right side.

Reply:
Ah I missed this, ty